### DISCUSSION TITLE
Grant user permissions in rootless container

### DISCUSSION THREAD

**Discussion by dmenneck** [containers/podman#19045] [General]
I try to run Rootless Podman as non-root user and get permission denied errors when starting a postgres container. The following commands and files are used:

```sh
podman build -t postgres -f containerfiles/Containerfile.postgres
```

**./containerfiles/Containerfile.postgres**
```yml
FROM docker.io/library/postgres:15.3

ENV POSTGRES_USER postgres
ENV POSTGRES_PASSWORD postgres
ENV POSTGRES_DB mapsolution_xq

COPY init.sql /docker-entrypoint-initdb.d/init.sql

EXPOSE 5432
```

When starting the container, I made sure to pass the `-u` flag to run the container as non-root:

```sh
podman run -u sync -d --name postgres postgres
```

The output of `podman logs postgres` is:

```sh
podman logs postgres
WARN[0000] Error validating CNI config file /home/ipuser/.config/cni/net.d/ipsysconnetwork.conflist: [plugin bridge does not support config version "1.0.0" plugin portmap does not support config version "1.0.0" plugin firewall does not support config version "1.0.0" plugin tuning does not support config version "1.0.0"]
chmod: changing permissions of '/var/lib/postgresql/data': Operation not permitted
chmod: changing permissions of '/var/run/postgresql': Operation not permitted
The files belonging to this database system will be owned by user "sync".
This user must also own the server process.

The database cluster will be initialized with locale "en_US.utf8".
The default database encoding has accordingly been set to "UTF8".
The default text search configuration will be set to "english".

Data page checksums are disabled.

fixing permissions on existing directory /var/lib/postgresql/data ... initdb: error: could not change permissions of directory "/var/lib/postgresql/data": Operation not permitted
```

I'm running podman on WSL. Is there anything i can change to fix the permission errors?

**Comment by rhatdan** [comment]
@n1hility PTAL

**Reply by dmenneck** [reply]
Hello @rhatdan and @n1hility, Unfortunately, I am still facing the same problem. As soon as I specify a volume, I get the following error message:

```sh
podman run --pod myPod -v ./postgres_data/:/var/lib/postgresql/data -d --name postgres postgres
```

```sh
chmod: /var/lib/postgresql/data: Operation not permitted
```

This error also occurs when I start the container as root (sudo) and pass the --privileged flag. It also doesn't matter whether I start the container in the WSL or in the Windows terminal. Do you have any ideas on how I can solve the problem?

**Reply by n1hility** [reply]
@dmenneck is this on the wsl default ubuntu distro? Are you in the current working directory on a windows drive? (e.g. when you type wsl it puts you in the CWD of your windows prompt by default). If on the WSL prompt you run `cd` to move into your linux home directory on WSL (on ext4), and then you run the podman command does it work?

**Reply by eriksjolund** [reply]
Does it work with the command-line option `--userns=keep-id:uid=999,gid=999`?

Longer story:
As a test I ran some popular container images  to check which ownership newly created directories and files have. See result: https://github.com/eriksjolund/podman-detect-option/tree/main#result-from-some-popular-container-images

The container image __docker.io/library/postgres__ produced some directories and files having UID=999 GID=999.

The command-line option `--userns=keep-id:uid=999,gid=999` maps
the regular user on host to the container user with UID=999 and GID=999.

I tried it out on Fedora CoreOS

### test1 (without --userns=keep-id:uid=999,gid=999)

```
[test@localhost ~]$ mkdir data1
[test@localhost ~]$ podman run --name test --rm -d \
                    -e POSTGRES_PASSWORD=insecurepassword \
                    -v ./data1:/var/lib/postgresql/data:Z \
                    docker.io/library/postgres
f4d82f477fb1a79ced1e971dc0808f987d1ef80a7c02b96198bab8e00ff2c7c3
[test@localhost ~]$ podman container stop test
[test@localhost ~]$ ls -l data1
ls: cannot open directory 'data1': Permission denied
[test@localhost ~]$ ls -ld data1
drwx------. 19 1442790 test 4096 Nov  9 19:14 data1
[test@localhost ~]$ ls -ldn data1
drwx------. 19 1442790 1014 4096 Nov  9 19:14 data1
[test@localhost ~]$ grep test /etc/subuid
test:1441792:65536
[test@localhost ~]$ id -u
1014
[test@localhost ~]$ id -un
test
[test@localhost ~]$ id -gn
test
```

__Result:__ the data directory is owned by UID=1442790 GID=1014.  This UID is one of the subuids belonging to the regular user on the host.

### test2 (with --userns=keep-id:uid=999,gid=999)

```
[test@localhost ~]$ mkdir data2
[test@localhost ~]$ podman run --name test --rm -d \
                    -e POSTGRES_PASSWORD=insecurepassword \
                    -v ./data2:/var/lib/postgresql/data:Z \
                    --userns=keep-id:uid=999,gid=999 \
                    docker.io/library/postgres
15ae09117bd90363d301f072950613b3bed2f73138194c47d7ffcde4f9965ed3
[test@localhost ~]$ podman container stop test
[test@localhost ~]$ ls -l data2 | head -5
total 64
-rw-------. 1 test test     3 Nov  9 19:01 PG_VERSION
drwx------. 5 test test    33 Nov  9 19:01 base
drwx------. 2 test test  4096 Nov  9 19:02 global
drwx------. 2 test test     6 Nov  9 19:01 pg_commit_ts
[test@localhost ~]$
```

__Result:__  the data directory is owned by the regular user on the host.

**Reply by dmenneck** [reply]
Thank you very much for your help! Your instruction...

```sh
mkdir data2

podman run --name test -d -e POSTGRES_PASSWORD=insecurepassword -v ./data2:/var/lib/postgresql/data:Z --userns=keep-id:uid=999,gid=999 docker.io/library/postgres
```

...results in (on Windows):

`podman logs test`
```sh
The files belonging to this database system will be owned by user "postgres".
This user must also own the server process.

The database cluster will be initialized with locale "en_US.utf8".
The default database encoding has accordingly been set to "UTF8".
The default text search configuration will be set to "english".

Data page checksums are disabled.

fixing permissions on existing directory /var/lib/postgresql/data ... ok
creating subdirectories ... ok
selecting dynamic shared memory implementation ... posix
selecting default max_connections ... 20
selecting default shared_buffers ... 400kB
selecting default time zone ... Etc/UTC
creating configuration files ... ok
2023-11-16 08:23:13.617 UTC [64] FATAL:  data directory "/var/lib/postgresql/data" has invalid permissions
2023-11-16 08:23:13.617 UTC [64] DETAIL:  Permissions should be u=rwx (0700) or u=rwx,g=rx (0750).
child process exited with exit code 1
initdb: removing contents of data directory "/var/lib/postgresql/data"
running bootstrap script ...
```

I would also like to start the Container within a pod. When I try to do this, the following error occurs:
```sh
Error: --userns and --pod cannot be set together
```

**Reply by eriksjolund** [reply]
Seeing this line
```
2023-11-16 08:23:13.617 UTC [64] DETAIL:  Permissions should be u=rwx (0700) or u=rwx,g=rx (0750).
```
Does it work if you run this command?
```
chmod 0700 data2
```

-----------

Regarding 

```
Error: --userns and --pod cannot be set together
```
If you need to use a pod then you need to instead specify the __--userns__ option to the __podman pod create__ command:

```
podman pod create --userns=keep-id:uid=999,gid=999 mypod
```
https://docs.podman.io/en/latest/markdown/podman-pod-create.1.html#userns-mode

----------------------

A side note: If you would need to run containers with different __--userns__ then it's not possible to run them in a pod.

**Reply by dmenneck** [reply]
The `Error: --userns and --pod cannot be put together` problem is now fixed, thanks! 

When I run `chmod 0700 data2` and restart the container, the following error message appears:

```sh
chmod: changing permissions of '/var/lib/postgresql/data': Operation not permitted
The files belonging to this database system will be owned by user "postgres".
This user must also own the server process.

The database cluster will be initialized with locale "en_US.utf8".
The default database encoding has accordingly been set to "UTF8".
The default text search configuration will be set to "english".

Data page checksums are disabled.

fixing permissions on existing directory /var/lib/postgresql/data ... initdb: error: could not change permissions of directory "/var/lib/postgresql/data": Operation not permitted
```

**Reply by eriksjolund** [reply]
> and restart the container

I think it will be easier to debug the problem if we  add the  __--rm__ option

```
mkdir data2
chmod 0700 data2
podman run --name test --rm -d -e POSTGRES_PASSWORD=insecurepassword -v ./data2:/var/lib/postgresql/data:Z --userns=keep-id:uid=999,gid=999 docker.io/library/postgres
```

I tried it out on a Fedora CoreOS machine.  It seems to have worked:
```
$ podman logs test 2>&1  | tail -1
2023-11-15 13:16:57.879 UTC [1] LOG:  database system is ready to accept connections
```

**Reply by eriksjolund** [reply]
#### summary

Maybe the podman you are using is rootful?

When running as root I also see the error message:
```
chmod: changing permissions of '/var/lib/postgresql/data': Operation not permitted
```
--------------

#### details

I tried to run the following commands as root on a Fedora CoreOS machine:


```
 mkdir data2
chmod 0700 data2
podman run --name test --rm -e POSTGRES_PASSWORD=insecurepassword -v ./data2:/var/lib/postgresql/data:Z --userns=keep-id:uid=999,gid=999 docker.io/library/postgres
```
(I didn't use `-d`)

This is the output from the commands:
```
chmod: changing permissions of '/var/lib/postgresql/data': Operation not permitted
The files belonging to this database system will be owned by user "postgres".
This user must also own the server process.

The database cluster will be initialized with locale "en_US.utf8".
The default database encoding has accordingly been set to "UTF8".
The default text search configuration will be set to "english".

Data page checksums are disabled.

initdb: error: could not access directory "/var/lib/postgresql/data": Permission denied
```

**Reply by dmenneck** [reply]
It could be that my podman is rootful, unfortunately I don't know. How can I 1) check if my podman is rootful and 2) how do I switch to rootless?

**Reply by eriksjolund** [reply]
To check if your podman is running rootful or rootless, run something like this

```
$  podman machine inspect testmachine | grep Rootful
          "Rootful": false
```

To see the list of machines:

```
$  podman machine list
```



If you would like to experiment, you could try
```
podman machine set --rootful=false machinename
```

When creating new machines there is also a __--rootful__ option
https://docs.podman.io/en/latest/markdown/podman-machine-init.1.html#rootful

**Reply by dmenneck** [reply]
I have created a new rootless (unprivileged) machine and unfortunately the same error is still occurring. Apparently I still don't have the rights to manage /var/lib/postgresql/data.

**Reply by eriksjolund** [reply]
Just to test some basic functionality, does this work?
```
mkdir testdir
podman run --rm -v ./testdir:/dir:Z  docker.io/library/alpine touch /dir/testfile
ls testdir/testfile
```

**Reply by eriksjolund** [reply]
Next time I get hold of a Windows computer that I could install Podman on, I could try to see if I could get it to work.
(Right now I only have access to Linux and macOS computers so it will probably take a while)

**Reply by dmenneck** [reply]
> Just to test some basic functionality, does this work?
> 
> ```
> mkdir testdir
> podman run --rm -v ./testdir:/dir:Z  docker.io/library/alpine touch /dir/testfile
> ls testdir/testfile
> ```

Yes, that works!

**Reply by eriksjolund** [reply]
I tried it out on macOS. There it works. I had to remove `:Z`.


```
$ podman machine start mymachine
$ podman system connection list | grep mymachine
mymachine                    ssh://core@127.0.0.1:60852/run/user/501/podman/podman.sock  /Users/myuser/.ssh/mymachine               false
mymachine-root               ssh://root@127.0.0.1:60852/run/podman/podman.sock           /Users/myuser/.ssh/mymachine               false
$
```
Note that I add `-c mymachine` to the podman commands below:
```
$ cd $HOME
$ mkdir data2
$ chmod 0700 data2
$ podman -c mymachine run --name test --rm -d -e POSTGRES_PASSWORD=insecurepassword -v ./data2:/var/lib/postgresql/data --userns=keep-id:uid=999,gid=999 docker.io/library/postgres  
5c5ec5a85e1b1e714309fb70cf936488f6bd657123ffa4193e448f7805a2e0d5
$ podman -c mymachine logs test 2>&1 | tail -1
2023-11-16 11:27:04.876 UTC [1] LOG:  database system is ready to accept connections
$
```

postgres has created files and directories in the bind-mounted directory:
```
$ ls -l data2 | head -4
total 120
-rw-------   1 myuser mygroup     3 Nov 16 12:26 PG_VERSION
drwx------   5 myuser mygroup   160 Nov 16 12:27 base
drwx------  65 myuser mygroup   2080 Nov 16 12:29 global
```

**Reply by dmenneck** [reply]
Is it possible that it does not work under Windows and in the WSL?

**Reply by eriksjolund** [reply]
I have never tried WSL so I don't know. Maybe someone else knows?

Is there any directory that can be accessed from both the host and from the VM?
If there is, create the _data2_ directory there.

**Reply by eriksjolund** [reply]
I realised that by adding `--user 0:0` to the __podman run__ command:

```
mkdir data2
chmod 0700 data2
podman run --user 0:0 --name test --rm -d -e POSTGRES_PASSWORD=insecurepassword -v ./data2:/var/lib/postgresql/data:Z --userns=keep-id:uid=999,gid=999 docker.io/library/postgres
```
the commands work on both rootless and rootful system connections.

**Reply by godefroi** [reply]
@eriksjolund It does not work for me on Windows+WSL:

```bash
podman run --user 0:0 --userns=keep-id:uid=999,gid=999 --interactive --rm -e "POSTGRES_PASSWORD=password" -v ./_postgres_data/:/var/lib/postgresql/data docker.io/library/postgres:17.6
```

Results in this:

```
The files belonging to this database system will be owned by user "postgres".
This user must also own the server process.

The database cluster will be initialized with locale "en_US.utf8".
The default database encoding has accordingly been set to "UTF8".
The default text search configuration will be set to "english".

Data page checksums are disabled.

fixing permissions on existing directory /var/lib/postgresql/data ... ok
creating subdirectories ... ok
selecting dynamic shared memory implementation ... posix
selecting default "max_connections" ... 25
selecting default "shared_buffers" ... 400kB
selecting default time zone ... Etc/UTC
creating configuration files ... ok
running bootstrap script ... 2026-02-05 20:28:19.068 UTC [77] FATAL:  data directory "/var/lib/postgresql/data" has invalid permissions
2026-02-05 20:28:19.068 UTC [77] DETAIL:  Permissions should be u=rwx (0700) or u=rwx,g=rx (0750).
child process exited with exit code 1
initdb: removing contents of data directory "/var/lib/postgresql/data"
```

Copilot suggests that uid, gid, and mode can be specified for the bind mount, but Podman complains if I attempt that. Copilot also suggests that enabling "wsl metadata" will solve the problem, but I have not (yet) tried that.

**Comment by dmenneck** [comment]
> @dmenneck is this on the wsl default ubuntu distro? Are you in the current working directory on a windows drive? (e.g. when you type wsl it puts you in the CWD of your windows prompt by default). If on the WSL prompt you run `cd` to move into your linux home directory on WSL (on ext4), and then you run the podman command does it work?

Yes! It works. 
However, the volume folder is not reflected outside wsl.

**Reply by n1hility** [reply]
@dmenneck  so provided the files are accessible to the wsl host user, for example using --userns=keep-id --user=1000, then you could access them from windows using `\\wsl$`. For example:  \\wsl$\podman-machine-default\home\user\postgres-data . Although what use-case are you thinking of with that? Backups? The reason I ask is that Postgres expects exclusive access to these.

**Reply by dmenneck** [reply]
We'd like to store all the data outside the container.
